ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java

```diff
@@ -295,15 +295,23 @@ private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size)
                 "Upload", partitionId.toPathString());
       }
       addToCache(blobId.getID(), (short) 0, BlobState.CREATED);
-      if (!uploaded) {
+      if (!uploaded && !isVcr) {
+        // If put is coming from frontend, then uploadBlob must be true. Its not acceptable that a blob already exists.
+        // If put is coming from vcr, then findMissingKeys might have reported a key to be missing even though the blob
+        // was uploaded to ABS. This can happen, if previously, the upload to ABS succeeded but the upload couldn't make
+        // it to cosmos.
         throw new StoreException(String.format("Another blob with same key %s exists in store", blobId.getID()),
             StoreErrorCodes.Already_Exist);
       }
     } else {
       vcrMetrics.blobUploadSkippedCount.inc();
-      throw new CloudStorageException("Error updating blob metadata", new StoreException(
-          String.format("Another blob with same key %s exists in store", messageInfo.getStoreKey().getID()),
-          StoreErrorCodes.Already_Exist));
+      // The only case where its ok to see a put request for a already seen blob is, during replication if the blob is
+      // expiring within {@link CloudConfig#vcrMinTtlDays} for vcr to upload.
+      if (isVcr && !isExpiringSoon(messageInfo)) {
```

Comment:
I don't understand this logic.  What about the case where we have reset the replica tokens but not deleted the blob store?  The put needs to be idempotent for VCR.

Reply:
I guess one way to look at this is that for vcr (replication), put happens when ReplicaThread has made the determination that this blob is not present locally by calling the CloudBlobStore::findMissingKeys(). So in case where we have reset the replica tokens but not deleted the blob store, a call to findMissingKeys should say that this blob is indeed present, and Replica Thread should not attempt a put() for this blob.

Let me know if I am missing something here.

Reply:
Okay, you may be right.